RV64 drop transmute usize<->ptr conversion

furquan-rivos opened the pull request.

adurbin-rivos commented on sys/kern/src/arch/rv64/task.rs:
```diff
@@ -34,7 +34,7 @@ pub unsafe fn get_current_task() -> &'static task::Task {
         asm!("csrr {0}, mscratch", out(reg) task);
         uassert!(task != 0);
-        core::mem::transmute::<usize, &task::Task>(task)
+        &*(task as *const task::Task)
```
Does this feel dirty? We're casting task as a const pointer to only deref and take address of it to obtain the pointer value. I know that's what has to be done, I assume obviously.  Not that the original was correct either. Makes me wonder if there's more to teach to the compiler on this one in the long run.

furquan-rivos replied:
> Does this feel dirty? We're casting task as a const pointer to only deref and take address of it to obtain the pointer value. I know that's what has to be done, I assume obviously.

yes, it feels really dirty. I am hoping there is some way to do this in a cleaner way and I am just not aware of it. I have added @picostove and @zachreizner to see if they have any inputs on doing this differently.

> Makes me wonder if there's more to teach to the compiler on this one in the long run.

That is probably what I concluded. But I could be missing something that Rust already provides.

zachreizner replied:
The canonical way to turn pointers into references in Rust is to use the weird looking `&*` operators. That's actually two operators (an unsafe Deref followed by a borrow) known as a reborrow: https://doc.rust-lang.org/reference/types/pointer.html#raw-pointers-const-and-mut.

tdewey-rivos reviewed (dismissed):
Approving on basis of bias for action, but cannot confirm if this is the Rust Way.

zachreizner commented:
The linked Rust doc on transmute only says there is unexpected behavior on pointer to integer. It's silent on the subject of integer to pointer. Perhaps only the pointer to integer side needs to be changed by this commit.

furquan-rivos commented:
> The linked Rust doc on transmute only says there is unexpected behavior on pointer to integer. It's silent on the subject of integer to pointer. Perhaps only the pointer to integer side needs to be changed by this commit.

The example [here](https://doc.rust-lang.org/core/mem/fn.transmute.html#examples) has the following comment: 
```
// Crucially, we `as`-cast to a raw pointer before `transmute`ing to a function pointer.
// This avoids an integer-to-pointer `transmute`, which can be problematic.
// Transmuting between raw pointers and function pointers (i.e., two pointer types) is fine.
```
So, I assumed that both conversions pointer->integer and integer->pointer are problematic.

furquan-rivos commented:
Looks like probably this change is the expected way to do things in Rust. I will wait until end of day and if no further comments I will go ahead and merge this. Thanks!

furquan-rivos committed:
kern/rv64: Drop usage of transmute for usize<->ptr

stove@ pointed out that use of transmute for usize<->ptr conversion
might not behave as expected. Also captured in the documentation of
transmute in
https://doc.rust-lang.org/core/mem/fn.transmute.html. This change
instead falls back to using `as` cast for the conversions.

TEST=Verified that rv64 qemu tests still pass.

furquan-rivos commented:
> [furquan-rivos](https://github.com/furquan-rivos) [force-pushed](https://github.com/rivosinc/hubris/compare/176c2d0124744ccdce389a403eea53d6c1691ee9..20722d54984d1a101a60fdc51d6ccf2b39d30dfe) the dev/furquan/drop-transmute branch from [176c2d0 ](https://github.com/rivosinc/hubris/commit/176c2d0124744ccdce389a403eea53d6c1691ee9)to [20722d5](https://github.com/rivosinc/hubris/commit/20722d54984d1a101a60fdc51d6ccf2b39d30dfe)

Looks like the conflict resolution I did on github didn't really work as I expected. So, I had to do it locally again and force push the dev branch. Need another review :-|.
